xarray/core/alignment.py

```diff
@@ -580,16 +580,21 @@ def reindex_variables(
 
     for name, var in variables.items():
         if name not in indexers:
+            if isinstance(fill_value, dict):
+                fill_value_ = fill_value.get(name, dtypes.NA)
```

Comment:
nice! this should provide a solution to https://github.com/pydata/xarray/issues/3390